mempool/validation: mempool ancestor/descendant limits for packages

glozow opened the pull request:
This PR implements a function to calculate mempool ancestors for a package and enforces ancestor/descendant limits on them as a whole. It reuses a portion of `CalculateMemPoolAncestors()`; there's also a small refactor to move the reused code into a generic helper function. Instead of calculating ancestors and descendants on every single transaction in the package and their ancestors, we use a "worst case" heuristic, treating every transaction in the package as each other's ancestor and descendant. This may overestimate everyone's counts, but is still pretty accurate in the our main package use cases, in which at least one of the transactions in the package is directly related to all the others (e.g. 1 parent + 1 child, multiple parents with 1 child, or chains).

Note on Terminology: While "package" is often used to describe groups of related transactions _within_ the mempool, here, I only use package to mean the group of not-in-mempool transactions we are currently validating.

#### Motivation

It would be a potential DoS vector to allow submission of packages to mempool without a proper guard for mempool ancestors/descendants. In general, the purpose of mempool ancestor/descendant limits is to limit the computational complexity of dealing with families during removals and additions. We want to be able to validate multiple transactions on top of the mempool, but also avoid these scenarios:

- We underestimate the ancestors/descendants during package validation and end up with extremely complex families in our mempool (potentially a DoS vector).
- We expend an unreasonable amount of resources calculating everyone's ancestors and descendants during package validation.

jnewbery commented:
Concept ACK. Treating every transaction in the package as each other's ancestor and descendant is a good, conservative heuristic to use, since it can never underestimate the true number of ancestors/descendants. If it's too limiting, we could potentially implement looser limits in future.

harding reviewed (commented):
Concept ACK.  The simplification taken of assuming each transaction has as many more ancestors and descendants as there are transactions in the package is a clever way to avoid writing/repurposing a bunch of code that's not necessary for the ultimate purpose of allowing package relay.

However, the more the results returned by TMPA diverge from the results we'd get from submitting each transaction to our mempool individually, the more I think package validation should be using a different interface than individual transaction validation (e.g., a different parameter to TMPA or a different RPC altogether).

Reviewed all the code, ran the functional tests, and tweaked the tests to see if I could get something unexpected to happen (nothing did).  I have one minor inline comment below about using a different error string when we're using different evaluation criteria.

I tried to think of additional tests, but the only thing I've got is that it might be nice to either replicate or convert the CPFP carve out tests from `test/functional/mempool_package_onemore.py` to use TMPA, e.g. as a variation of the current "A-shaped test" (I actually tried to do this real quick but couldn't get it to work; it was clearly my fault as I got the same error whether I used `submitrawtransaction` or TMPA).

Overall, this PR LGTM.  Thanks!

glozow commented:
@harding thank you for the review!!!

> However, the more the results returned by TMPA diverge from the results we'd get from submitting each transaction to our mempool individually, the more I think package validation should be using a different interface than individual transaction validation (e.g., a different parameter to TMPA or a different RPC altogether).

I agree, and they will likely continue to diverge if we add `bypass_timelocks` and such... Perhaps we can have a regtest-only rawpackage RPC with a `test_accept` parameter (in my opinion users should never have to interact with packages)? And testmempoolaccept can be for users / L2 testing?

harding commented:
@glozow 

> Perhaps we can have a regtest-only rawpackage RPC with a `test_accept` parameter (in my opinion users should never have to interact with packages)?

There's certainly no need for users to interact with packages before there's a reliable package relay mechanism, so starting with regtest-only seems like a good idea to me.  If someone later comes along with a reason it should be user-facing, they can put in the (probably trivial) amount of work to make the RPC available on mainnet and the other networks.

>  And testmempoolaccept can be for users / L2 testing?

For users, yes.  I don't think anyone is using it today for L2 testing and I'm not sure it's really well suited to that---testmempoolaccept tells you whether your transaction would be accepted into the current mempool, but L2 testers really want to know whether the transaction will be accepted into a future mempool; a failure now is a reliable harbinger of failure later, but a success now doesn't guarantee success in the future (even ignoring that relay policy can be made more restrictive).

michaelfolkson commented:
> For users, yes. I don't think anyone is using it today for L2 testing and I'm not sure it's really well suited to that---testmempoolaccept tells you whether your transaction would be accepted into the current mempool, but L2 testers really want to know whether the transaction will be accepted into a future mempool; a failure now is a reliable harbinger of failure later, but a success now doesn't guarantee success in the future (even ignoring that relay policy can be made more restrictive).

I'm not convinced this is true. It is impossible for L2 testers to know whether the transaction will be accepted into a future mempool as fee levels could theoretically be anything up to infinite. So there is nothing to test. So the L2 testing would be for checking that the package would be accepted at the current fee levels (e.g. possibly just before broadcasting the package or for general testing at current fee levels)

murchandamus reviewed (commented):
Concept ACK

glozow commented on src/txmempool.cpp:
```diff
@@ -187,10 +201,13 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
             return false;
-        } else if (stageit->GetCountWithDescendants() + 1 > limitDescendantCount) {
+        } else if (stageit->GetCountWithDescendants() + total_count > limitDescendantCount) {
             errString = strprintf("too many descendants for tx %s [limit: %u]", stageit->GetTx().GetHash().ToString(), limitDescendantCount);
```
Added a "possibly" to the beginning of the error strings, gated on `total_count > 1`, so that the existing tests that assert expected error messages don't need to be changed.

glozow commented on src/txmempool.cpp:
```diff
@@ -151,33 +152,46 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
+                if (piter) {
+                    staged_ancestors.insert(**piter);
+                    if (staged_ancestors.size() + total_count > limitAncestorCount) {
+                        errString = strprintf("too many unconfirmed parents [limit: %u]", limitAncestorCount);
```
Done, see https://github.com/bitcoin/bitcoin/pull/21800#discussion_r649073219

glozow commented:
Incorporated doc suggestions from #22084 (to make it mergeable) and some review comments here, working on adding more edge casey tests that were discussed in PR Review Club last night.

glozow commented:
Rebase a837686...1061cf4

rajarshimaitra reviewed (commented):
tACK https://github.com/bitcoin/bitcoin/pull/21800/commits/1061cf457b021ccf24f394dbb5cc00af575598ba

Overall the idea of overestimating package ancestor and descendant seems reasonable to me. 

Stepped through the functional test, covering A and V shaped mempool ancestry. All tests passing.

glozow commented on src/policy/packages.h:
```diff
@@ -14,6 +15,7 @@
 static constexpr uint32_t MAX_PACKAGE_COUNT{25};
 /** Default maximum total virtual size of transactions in a package in KvB. */
 static constexpr uint32_t MAX_PACKAGE_SIZE{101};
+static_assert(MAX_PACKAGE_SIZE * 4 * 1000 >= MAX_STANDARD_TX_WEIGHT);
```
You may find #22097 of interest :)

ariard reviewed (commented):
I think this PR around mempool ancestors/descendants limits illustrates well current limitations of our approach with implementing package-relay, namely the uncertainty about types of packages we want to support.

Union-based evaluation of ancestors/descendants limits doesn't work well as an API from a L2 protocol dev viewpoint. Let's say you implement [domino bumping](https://gnusha.org/lightning-dev/2021-05-14.log) to close in emergency multiple lightning channels, where the bumping utxo is reused across a CPFP chain:

```
        A         E         I
          \         \         \
           B -- D -- F -- H -- J -- K
               /        /          /
             C        G          L 
```

Where root vertices are commitment transactions.

If you already have {A,B,C,D} in the mempools, and try to broadcast {E,F,G,H,I,J,K,L) in a single package it's going to be rejected (4 * 8 > DEFAULT_ANCESTOR_LIMITS), though if you broadcast components by pair, it will be accepted.

Also, to the best of my knowledge, all deployed or in deployment L2s (LN, Revault, Lightning Pool) have the shared need of 2-sized package but not beyond. Honestly deploying a 2-txn sized package relay to solve the safety-critical issue about pre-signed feerate transactions under mempool min feerate would already be a huge win. Once done, we can consider loosening package API, as there is a need emerging in the ecosystem for other types of packages.

Further, from a base-layer viewpoint, once we have p2p packages, it doesn't seem reasonable to allow mempool traversals for chain of transactions with 25 elements. It's far easier to reason about DoS concerns with 2-txn sized packages.

So what do you think about ?

* screwing up MAX_PACKAGE_COUNT to 2 transactions for the initial deployement of package-relay
* implementing "intersection-dedup" evaluation, where duplicated ancestors/descendants aren't accounted for package limits ?

I think code changes should be limited as you can account the second package element as one-more descendant and substract it, including its size, to the `CalculateMemPoolAncestors`'s `limitDescendant*` argument of the first package element ?

ariard commented:
> I agree, and they will likely continue to diverge if we add bypass_timelocks and such... Perhaps we can have a regtest-only rawpackage RPC with a test_accept parameter (in my opinion users should never have to interact with packages)? And testmempoolaccept can be for users / L2 testing?

Sounds a good idea to introduce a `rawpackage` test-only RPC. I believe switching it to mainet is going to be function of which p2p package version we decide on (either sender-initiated or relay-initiated). W.r.t to `bypass_timelocks`/`bypass_feerate` I think that's ultimately different from the package interface as you can interested by the mempool acceptance of individual transactions, we can leave them on `testmempoolaccept` for now ?

@michaelfolkson 

> I'm not convinced this is true. It is impossible for L2 testers to know whether the package would be accepted into a future mempool as fee levels could theoretically be anything up to infinite. So there is nothing to test. The L2 testing would be for checking that the package would be accepted at the current fee levels (e.g. possibly just before broadcasting the package or for general testing at current fee levels)

Your point is valid, but note in #20833 we had the discussion to introduce `bypass_feerate` to remedy this pitfall (somewhere in the GH comments, yeah I know it's messy...)

glozow commented on src/validation.cpp:
```diff
@@ -1114,6 +1114,31 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
     }
 
+    // Limit the scope of _entries and _ancestors. We should calculate ancestors for each
+    // transaction individually when we call Finalize().
```
I should clarify this comment - I mean when we call `addUnchecked` in real package mempool accept, we'll need to recalculate the ancestors for each one as input

glozow replied:
fixed

glozow commented on src/txmempool.h:
```diff
@@ -681,6 +681,29 @@ class CTxMemPool
     bool CalculateMemPoolAncestors(const CTxMemPoolEntry& entry, setEntries& setAncestors, uint64_t limitAncestorCount, uint64_t limitAncestorSize, uint64_t limitDescendantCount, uint64_t limitDescendantSize, std::string& errString, bool fSearchForParents = true) const EXCLUSIVE_LOCKS_REQUIRED(cs);
 
+    /** Try to calculate all in-mempool ancestors of a set of entries, and check for ancestor and
+     * descendant limits (all are inclusive of the transactions in entries). The same limits are
```
It's documenting the way we apply the limits, e.g. if `ancestorcount` is 25, it means the total number of ancestors, including itself, must be within 25.

glozow commented on src/validation.cpp:
```diff
@@ -1114,6 +1114,31 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+        std::transform(workspaces.cbegin(), workspaces.cend(), std::back_inserter(_entries),
+                       [](const auto& ws) { return std::cref(*ws.m_entry); });
+        // We won't use the set of ancestors returned for calling Finalize().
+        CTxMemPool::setEntries _ancestors;
```
done

glozow commented on src/validation.cpp:
```diff
@@ -1114,6 +1114,31 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+                ws.m_state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-long-mempool-chain", err_string);
+                results.emplace(ws.m_ptx->GetWitnessHash(), MempoolAcceptResult::Failure(ws.m_state));
+            }
+            package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-too-long-mempool-chain");
```
changed the tx error to `exceeds-ancestor-descendant-limits` and the package error `package-mempool-limits`

glozow commented on src/txmempool.h:
```diff
@@ -681,6 +681,29 @@ class CTxMemPool
+     * @param[in]       fSearchForParents       Whether to search for entries' in-mempool parents.
+     *                                          Must be true if any entries are not already in mempool.
+     */
+    bool CalculateMemPoolAncestors(const std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef>& entries,
```
Added doxygen comment for `entries` param, but my intention was to overload the `CalculateMemPoolAncestors` function (the fact that the first argument is a vector instead of a single entry should suffice?)

glozow commented:
@ariard thanks for the review, I'm very much looking forward to the IRC discussions to better understand what packages we want to support. In your domino bumping example, I'm not entirely sure why we couldn't use several packages of size 2 to bump them?

> MAX_PACKAGE_COUNT to 2 transactions for the initial deployement of package-relay

We're still a few PRs from package accept with submission to mempool - will put some things up soon - and then a few PRs from package relay but sounds good to me. I personally believe that the P2P requirements can be more stringent than individual transaction mempool policy - I also think it's fine for peers to negotiate their maximum package size they're willing to receive in a `SENDPACKAGES` message. Starting with a default of 2 sounds good to me. However, I don't think package ancestor/descendant limits should be explicitly less than individual ones within a node, because I think it could give an adversary the chance to prevent a package from being accepted by submitting a very large descendant to the top transaction.

> implementing "intersection-dedup" evaluation, where duplicated ancestors/descendants aren't accounted for package limits ?

I'm not sure I understand, could you elaborate? The `setAncestors` used in `CalculateMemPoolAncestors()` is not duplicated between transactions. I am working on a branch to "trim" packages when it overlaps with mempool (i.e. one or more transactions are already in the mempool), and will add test cases for ancestor/descendant limits with it.

(Recent push addresses inline review comments)

Xekyo commented:
> If you already have {A,B,C,D} in the mempools, and try to broadcast {E,F,G,H,I,J,K,L) in a single package it's going to be rejected (4 * 8 > DEFAULT_ANCESTOR_LIMITS), though if you broadcast components by pair, it will be accepted.

@ariard: I don't understand the reasoning that leads to "4 * 8" here.

glozow commented:
Bump, this is the next step for package validation logic.

The heuristic used in this implementation still results in an exact calculation of in-mempool and in-package ancestors for packages of parent + child (the MVP/basic package use case as discussed in the IRC meetings). It can be slightly more restrictive in scenarios such as batched fee bumping, but is extensible if we want to implement something more granular in the future.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -151,33 +152,47 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
+        // in the mempool.
+        assert(entries.size() == 1);
+        txiter it = mapTx.iterator_to(entries[0].get());
+        staged_ancestors = it->GetMemPoolParents();
```
What happened to the `Const` here? This used to be `GetMemPoolParentsConst()`.

glozow replied:
Fixed

jnewbery commented on src/txmempool.cpp:
```diff
@@ -151,33 +152,47 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
 }
 
-bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntries &setAncestors, uint64_t limitAncestorCount, uint64_t limitAncestorSize, uint64_t limitDescendantCount, uint64_t limitDescendantSize, std::string &errString, bool fSearchForParents /* = true */) const
+bool CTxMemPool::CalculateMemPoolAncestors(const std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef>& entries,
```
I think that `CalculateMemPoolAncestors()` was already doing too much, and this adds just a little bit more complication. Instead of overloading CMPA and having CMPA(tx) call CMPA(package), what do you think about the splitting it into two parts:

- calculate parents (two different functions - one for a single transaction which may have fSearchForParents=false, and one for a package)
- check package limits (everything from `while (!staged_ancestors.empty())` downwards, which is the same function for an individual transaction and a package if parametrized correctly)

and then roughly:

```
CMPA(tx) {
    set_ancestors = CalculateParents(tx);
    CheckPackageLimits(set_ancestors);
}

CMPA(package) {
    set_ancestors = CalculateParents(package);
    CheckPackageLimits(set_ancestors);
}
```

JeremyRubin replied:
nit: use function signatures with spans when passing a const vec, removes the need to allocate for passing a single tx (can span without allocating from pointer).

JeremyRubin replied:
re: @jnewbery's feedback to split, I think that might be right if fsearchforparents should never be used as false with a package? the asserts confuse me...

jnewbery commented on src/txmempool.cpp:
```diff
@@ -151,33 +152,47 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
-bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntries &setAncestors, uint64_t limitAncestorCount, uint64_t limitAncestorSize, uint64_t limitDescendantCount, uint64_t limitDescendantSize, std::string &errString, bool fSearchForParents /* = true */) const
+bool CTxMemPool::CalculateMemPoolAncestors(const std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef>& entries,
+                                           setEntries& setAncestors,
+                                           const uint64_t limitAncestorCount,
+                                           const uint64_t limitAncestorSize,
+                                           const uint64_t limitDescendantCount,
+                                           const uint64_t limitDescendantSize,
```
What do you think about not making these const, and then decrementing from them and checking that we don't go below zero? That would avoid the need for the local variables and make the logic for individual txs and packages more similar.

JeremyRubin replied:
I think it is clearer to make these as const personally, decrement until 0 is always confusing to me (is it at 0 or below zero that the error comes in? These are currently uint64_t, so we'd also need to either convert to int64_t or detect wraparound...).

glozow replied:
Another problem with editing these variables is we'd need to change the error logging (current log would use the decremented count instead of actual limit), so I've left it as is.

JeremyRubin commented on src/txmempool.cpp:
```diff
@@ -151,33 +152,47 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
-        txiter it = mapTx.iterator_to(entry);
-        staged_ancestors = it->GetMemPoolParentsConst();
+        // If we're not searching for parents, we require all entries to be in the mempool already.
+        // We should never use CalculateMemPoolAncestors on a set of transactions that are already
```
Never? Why is this new constraint present? I thought it's safe to use as long as we aren't searching for parents?

glozow replied:
(Same idea) if we're calling this with multiple transactions, they shouldn't already be in the mempool.

glozow replied:
Removed this constraint

JeremyRubin commented on src/txmempool.cpp:
```diff
@@ -151,33 +152,47 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
+        // If we're not searching for parents, we require all entries to be in the mempool already.
+        // We should never use CalculateMemPoolAncestors on a set of transactions that are already
+        // in the mempool.
+        assert(entries.size() == 1);
```
this assert is pretty scary? why do we not permit >1?

JeremyRubin replied:
What if it's a package of size 1? Is this code safe? Or only if it's called in non package contexts?

glozow replied:
Sorry, I've clarified this now. CMPA for a vector of transactions always searches for parents. If we call CMPA with multiple transactions (also handling the case if called with a vector of size 1), those transactions shouldn't already be in the mempool.

glozow replied:
Removed this condition

JeremyRubin commented on src/txmempool.cpp:
```diff
@@ -216,6 +240,17 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+                                           uint64_t limitDescendantCount, uint64_t limitDescendantSize,
+                                           std::string &errString, bool fSearchForParents /* = true */) const
+{
+    std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef> entry_vec{std::cref(entry)};
```
if you update to a span, you get rid of an allocation here!

glozow replied:
GONE! 👐

glozow commented on src/txmempool.cpp:
```diff
@@ -187,14 +202,22 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+        // stricter than the limits for each individual transaction.
+        if (stageit->GetSizeWithDescendants() + total_virtual_size > limitDescendantSize) {
+            errString = strprintf("%sexceeds descendant size limit for tx %s [limit: %u]",
+                                  total_count > 1 ? "possibly" : "",
```
Fixed

JeremyRubin replied:
did you tho? looks same

glozow commented on test/functional/rpc_packages.py:
```diff
@@ -187,6 +190,178 @@ def test_chain(self):
+        rawtx = node.createrawtransaction(inputs, outputs)
+
+        parent_signed = node.signrawtransactionwithkey(hexstring=rawtx, privkeys=self.privkeys)
+        parent_tx = CTransaction()
```
Fixed

glozow commented on src/txmempool.h:
```diff
@@ -681,6 +681,34 @@ class CTxMemPool
+     *                                          They need not be direct ancestors/descendants of
+     *                                          each other, though they will be treated as such.
+     * @param[in,out]   setAncestors            Set of in-mempool ancestors. Updated to include
+     *                                          any new ancestors found.
```
You're right it's just an out param, fixed.

glozow commented on src/txmempool.h:
```diff
@@ -681,6 +681,34 @@ class CTxMemPool
+     * descendant limits (all are inclusive of the transactions in entries). The same limits are
+     * used no matter how many transactions are passed in. For example, if entries.size() = 3 and
+     * the limit is 25, the union of all 3 sets of ancestors must be <= 22.
+     * @param[in]       entries                 Entries corresponding to transaction(s) being
+     *                                          evaluated for acceptance to mempool. If there are
```
Took it out. Also kind of different now that the two CMPAs are slightly different

glozow committed:
MOVEONLY: add helper function for calculating ancestors and checking limits

glozow committed:
[refactor] pass size/count instead of entry to CalculateAncestorsAndCheckLimits

This does not change existing behavior.
The ancestor/descendant limits are inclusive of the entries themselves,
but CalculateAncestorsAndCheckLimits() does not need access to them.

glozow committed:
misc package validation doc improvements

glozow committed:
[mempool] check ancestor/descendant limits for packages

When calculating ancestor/descendant counts for transactions in the
package, as a heuristic, count every transaction in the package as an
ancestor and descendant of every other transaction in the package.

This may overestimate, but will not underestimate, the
ancestor/descendant counts. This shortcut still produces an accurate
count for packages of 1 parent + 1 child.

glozow committed:
[policy] ancestor/descendant limits for packages

glozow committed:
[test] mempool package ancestor/descendant limits

glozow commented on src/txmempool.h:
```diff
@@ -681,6 +681,34 @@ class CTxMemPool
+     * the limit is 25, the union of all 3 sets of ancestors must be <= 22.
+     * @param[in]       entries                 Entries corresponding to transaction(s) being
+     *                                          evaluated for acceptance to mempool. If there are
+     *                                          multiple, they must not already be in the mempool.
```
Right, removed

jnewbery commented on src/txmempool.h:
```diff
@@ -585,6 +586,25 @@ class CTxMemPool
 
+
+    /**
+     * Helper function to calculate all in-mempool ancestors of staged_ancestors and apply
```
```suggestion
     * Helper function to calculate all in-mempool ancestors of staged_ancestors (including staged_ancestors themselves) and apply
```

or similar to document that setAncestors will include the transactions in staged_ancestors.

glozow replied:
Done

jnewbery commented on src/txmempool.h:
```diff
@@ -585,6 +586,25 @@ class CTxMemPool
+     */
+    bool calculateAncestorsAndCheckLimits(size_t entry_size,
+                                          size_t entry_count,
+                                          setEntries& setAncestors,
```
The `&` jumps from right to left between commits. Perhaps just have this as `setEntries& setAncestors` in commit _MOVEONLY: add helper function for calculating ancestors and checking limits_

glozow replied:
Done

jnewbery commented on src/txmempool.cpp:
```diff
@@ -187,14 +170,20 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
-            errString = strprintf("exceeds descendant size limit for tx %s [limit: %u]", stageit->GetTx().GetHash().ToString(), limitDescendantSize);
+        if (stageit->GetSizeWithDescendants() + entry_size > limitDescendantSize) {
+            errString = strprintf("%sexceeds descendant size limit for tx %s [limit: %u]",
+                                  entry_count > 1 ? "possibly " : "",
```
These conditional log changes make the function less generic (it now knows something semantic about the caller). If you really want this "possibly " prefix, you could prepend it in the calling function:

```diff
     // When multiple transactions are passed in, the ancestors and descendants of all transactions
     // considered together must be within limits even if they are not interdependent. This may be
     // stricter than the limits for each individual transaction.
-    return calculateAncestorsAndCheckLimits(total_size, package.size(),
-                                            setAncestors, staged_ancestors,
-                                            limitAncestorCount, limitAncestorSize,
-                                            limitDescendantCount, limitDescendantSize, errString);
+    auto ret = calculateAncestorsAndCheckLimits(total_size, package.size(),
+                                                setAncestors, staged_ancestors,
+                                                limitAncestorCount, limitAncestorSize,
+                                                limitDescendantCount, limitDescendantSize, errString);
+
+    errString.insert(0, "possibly ");
+    return ret;
 }
```

Although I don't think you really need it. If someone is submitting packages through the RPC, then presumably they're a sophisticated user and they'll be able to understand that a heuristic is being used to calculate ancestor/descendant limits.

You could also update the only place that this string is used:

```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1094,7 +1094,7 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
             // be implementation-dependent, and it's likely to be multiple transactions because we
             // evaluated all of them together. Return the same failure for all transactions.
             for (auto& ws : workspaces) {
-                ws.m_state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "exceeds-ancestor-descendant-limits", err_string);
+                ws.m_state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "exceeds-ancestor-descendant-limits-heuristic", err_string);
                 results.emplace(ws.m_ptx->GetWitnessHash(), MempoolAcceptResult::Failure(ws.m_state));
             }
             package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-mempool-limits");
```

glozow replied:
Good point, added the "possibly " prefix outside

jnewbery commented on src/txmempool.cpp:
```diff
@@ -216,6 +207,69 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+        staged_ancestors = it->GetMemPoolParentsConst();
+    }
+
+    return calculateAncestorsAndCheckLimits(entry.GetTxSize(), 1, setAncestors, staged_ancestors,
```
Document basic type arguments:

```suggestion
    return calculateAncestorsAndCheckLimits(entry.GetTxSize(), /* entry_count */ 1, setAncestors, staged_ancestors,
```

glozow replied:
Added

jnewbery commented on src/txmempool.cpp:
```diff
@@ -216,6 +207,69 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+                    errString = strprintf("too many unconfirmed parents [limit: %u]", limitAncestorCount);
+                    return false;
+                }
+            }
```
Would it make sense to have an early exit here if `total_size` exceeds `limitAncestorSize` (similar to if the ancestor count exceeds `limitAncestorCount`)?

glozow replied:
Not sure. In practice, it wouldn't happen since we would have checked the package limits already

jnewbery commented on src/txmempool.cpp:
```diff
@@ -216,6 +207,69 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
     return true;
 }
 
+bool CTxMemPool::CalculateMemPoolAncestors(const Package& package,
```
This function doesn't need to have a `setAncestors` out param, since the only place that it's called doesn't use it. In fact, you could rename this function to something like `CheckPackageLimits()` to better document what it's used for.

glozow replied:
Good point - done :)

jnewbery reviewed (commented):
This looks great. The tests are excellent.

I have a few suggestions inline.

ritickgoenka commented:
tACK 

Ran the tests locally and also tried to tweak the tests to make sure everything was working fine. 

While checking for ancestors limit for a transaction we are checking the total number of ancestors including itself should be within 25, but when we are checking for descendants limit for a transaction we are checking the total number of descendants excluding itself should be within 25. Is this the expected behavior? Or do we want to include the transaction also while checking the descendant limits?

To test that the transaction itself is excluded when checking descendants limit I changed the 'A' shaped test a little, I reduced the number of in mempool transactions in the left limb from 12 to 11 (total 26 transactions, 24 in mempool and 2 in package) and kept everything else the same and the test failed.

Similarly, to test that the transaction itself is included when checking ancestor limit I changed the 'V' shaped test, I reduced the number of in mempool transactions in the left limb from 12 to 11 (total 26 transactions, 23 in mempool and 3 in package) and kept everything else the same and the test Passed.


Tested 27986f8 on Ubuntu 18.04

glozow commented:
Thanks for the review @ritickgoenka!

> While checking for ancestors limit for a transaction we are checking the total number of ancestors including itself should be within 25, but when we are checking for descendants limit for a transaction we are checking the total number of descendants excluding itself should be within 25. 

I don't think this is true. Descendant limits are also inclusive of the transaction.

> To test that the transaction itself is excluded when checking descendants limit I changed the 'A' shaped test a little, I reduced the number of in mempool transactions in the left limb from 12 to 11 (total 26 transactions, 24 in mempool and 2 in package) and kept everything else the same and the test failed.

I looked into this, and realized that I mislabeled the transactions in the diagram - my bad. The current test has 24 transactions in the mempool and 2 in the package, so if you reduced the mempool transactions, it's 23+2 which is within limits. That's why the test would fail. I've updated the comment and added assertions for the number of transactions in mempool and package.

glozow commented:
Addressed @jnewbery's comments and added another test

glozow commented on src/txmempool.cpp:
```diff
@@ -151,34 +151,17 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
 }
 
-bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntries &setAncestors, uint64_t limitAncestorCount, uint64_t limitAncestorSize, uint64_t limitDescendantCount, uint64_t limitDescendantSize, std::string &errString, bool fSearchForParents /* = true */) const
+bool CTxMemPool::calculateAncestorsAndCheckLimits(size_t entry_size,
```
Done

jnewbery commented on test/functional/rpc_packages.py:
```diff
@@ -181,6 +185,232 @@ def test_chain(self):
+        # Clear mempool and check that the package passes now
+        node.generate(1)
+        assert all([res["allowed"] for res in node.testmempoolaccept(rawtxs=package_hex)])
+        node.generate(1)
```
No need for these new calls to `generate(1)`.

glozow replied:
Good point :P removed. Added asserts for empty mempool in each of the subtests.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -216,6 +200,79 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+                                                      setAncestors, staged_ancestors,
+                                                      limitAncestorCount, limitAncestorSize,
+                                                      limitDescendantCount, limitDescendantSize, errString);
+    if (!ret) errString.insert(0, "possibly ");
```
Might be worth commenting why you're prefixing the error string with "possibly" here.

fanquake replied:
Agree

glozow replied:
Done

laanwj replied:
It's somewhat surprising to have an existing error message prefixed with 'possibly', I checked that it works for the current messages in `CalculateAncestorsAndCheckLimits`, but the validity of this text manipulation seems to be hard to guarantee over future changes to the code.
But I don't know a better solution to this, The grammatical correctness of the error message is not super critical.

jnewbery reviewed (commented):
utACK 6d8f687bfc

New test looks good

ritickgoenka commented:
reACK 6d8f687

Tested the new test which was added. 
Also, wrote a new test to check the overestimation of descendant limits and it was working fine.

Tested on Ubuntu 18.04

tryphe commented:
Concept ACK

Really nice work!

fanquake reviewed (commented):
Concept ACK - the changes here look fairly straight forward. Need to properly review the test.

glozow committed:
extract/rename helper functions from rpc_packages.py

MOVEONLY; no change in behavior. Rename because there is another helper
funciton in chain_transaction in test_framework.util.py

ariard commented on src/txmempool.cpp:
```diff
@@ -187,10 +171,10 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+        } else if (stageit->GetCountWithDescendants() + entry_count > limitDescendantCount) {
             errString = strprintf("too many descendants for tx %s [limit: %u]", stageit->GetTx().GetHash().ToString(), limitDescendantCount);
             return false;
         } else if (totalSizeWithAncestors > limitAncestorSize) {
```
I think we have topologies of in-mempool transactions and packages bypassing this limit.

Let's say you have in-mempool, independent Tx_A and Tx_B where their virtual sizes are 30 KvB each.
Let's say you have packages transactions Tx_C and _Tx_D where Tx_C is a child of A and B and Tx_D is a child of Tx_C their virtual sizes are 30 KvB each.

In `CheckPackageLimits`, Tx_C isn't already in the mempool (`mapTx` is only updated in `Finalize`) and as such shouldn't part of `staged_ancestors`. This set should be composed of Tx_A and Tx_B only.

In `CalculateAncestorsAndCheckLimits`, `totalSizeWithAncestors` is the union Tx_A and Tx_B and its sum of 60 KvB is inferior to 101 KvB. After Tx_C and Tx_D are added to the mempool, the chain of transactions ABCD is of the sum 120 KvB.

Further, I don't think it's rejected by the check against `limitDescendantSize`, as Tx_A and Tx_B are evaluated independently, and not as a single cluster as they should be (Tx_A + Tx_C + Tx_D or Tx_B + Tx_C + Tx_D)

Note, you need to replace Tx by chain of transactions because of `MAX_STANDARD_TX_WEIGHT` but I think this reasoning holds ?

ariard replied:
Thanks for writing the test, I noticed where my reasoning was flawed! 

>  totalSizeWithAncestors is the union Tx_A and Tx_B and its sum of 60 KvB is inferior to 101 KvB. 

We init `totalSizeWithAncestors` with the whole package size from now, instead of the entry only (L164 in `src/txmempool.cpp`). So we have A+B+C+D  and the check against `limitAncestorSize` rejects the package. Maybe the variable could be renamed `packageSizeWithAncestors`, but that's a nit.

glozow commented on src/txmempool.cpp:
```diff
@@ -216,6 +200,79 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+            std::optional<txiter> piter = GetIter(input.prevout.hash);
+            if (piter) {
+                staged_ancestors.insert(**piter);
+                if (staged_ancestors.size() + package.size() > limitAncestorCount) {
```
Yep, this is absolutely a consideration for replacements and duplicates in packages in the future. I agree, we'll want to de-duplicate first (CMPA and CheckPackageLimits wouldn't be called for transactions already in the mempool, since PreChecks looks for `txn-already-in-mempool` before this). Additionally, we'll want to subtract the size/count of transactions being replaced from descendant limits before calling this function.

glozow replied:
(I will note this down for the future - I think we're on the same page about this, and it's not yet applicable for this PR)

ariard replied:
> Additionally, we'll want to subtract the size/count of transactions being replaced from descendant limits before calling this function.

Exactly, this is another limit case to be aware of. And even trickier one like an ancestor of a second-stage package member being replaced by a first-stage package member and thus failing the whole package acceptance.

Yes, not yet applicable for this PR, good to not forget about it!

JeremyRubin replied:
i think a simpler API would be to have staged always contain all the entries themselves? is there a reason not to (I think the epoch algorithm is relatively lightweight for this).

glozow commented on src/validation.cpp:
```diff
@@ -1084,6 +1084,22 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+        // evaluated all of them together. Return the same failure for all transactions.
+        for (auto& ws : workspaces) {
+            ws.m_state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "exceeds-ancestor-descendant-limits", err_string);
+            results.emplace(ws.m_ptx->GetWitnessHash(), MempoolAcceptResult::Failure(ws.m_state));
```
That's a good point - perhaps it would be better to copy the error string from the `TxValidationState` into the `PackageValidationState` and always return results empty when we have a package-wide error.

glozow replied:
Thanks for the suggestion - I got rid of the tx results when there's a package-wide error, and added a comment for `m_tx_results` in validation.h

glozow commented on src/txmempool.cpp:
```diff
@@ -187,10 +171,10 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
-        if (stageit->GetSizeWithDescendants() + entry.GetTxSize() > limitDescendantSize) {
+        if (stageit->GetSizeWithDescendants() + entry_size > limitDescendantSize) {
             errString = strprintf("exceeds descendant size limit for tx %s [limit: %u]", stageit->GetTx().GetHash().ToString(), limitDescendantSize);
             return false;
```
This error should be caught in the mempool_tests unit test. It's true, though, that I didn't write tests for the size limits, just did count limits.

glozow replied:
Added a functional test for size limits - `test_desc_size_limits()` will fail if you apply this diff

ariard replied:
> This error should be caught in the mempool_tests unit test?

It should but I don't' get an error for any `mempool_*` unit tests. Won't be surprised there is a hole coverage before to this PR. 

> Added a functional test for size limits - test_desc_size_limits() will fail if you apply this diff

Thanks for adding one, there is at least one behavior change to cover introduce with this PR, `entry_size` can be > 1.

glozow committed:
[test] helper function to increase transaction weight

glozow committed:
[test] parameterizable fee for make_chain and create_child_with_parents

laanwj commented:
Code review ACK 5d513c0698dfe088e6dcd77ade59a0b34d92efbd

fanquake commented:
https://cirrus-ci.com/task/6628050143019008:
```bash
Run tx_pool_standard with args ['/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz', '-runs=1', '/tmp/cirrus-ci-build/ci/scratch/qa-assets/fuzz_seed_corpus/tx_pool_standard']INFO: Seed: 517370066
INFO: Loaded 1 modules   (547316 inline 8-bit counters): 547316 [0x556b20760ed8, 0x556b207e68cc), 
INFO: Loaded 1 PC tables (547316 PCs): 547316 [0x556b207e68d0,0x556b21040810), 
INFO:     3949 files found in /tmp/cirrus-ci-build/ci/scratch/qa-assets/fuzz_seed_corpus/tx_pool_standard
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 3949 min: 1b max: 1048576b total: 95048838b rss: 213Mb
#128	pulse  cov: 17131 ft: 25503 corp: 122/895b exec/s: 64 rss: 225Mb
#256	pulse  cov: 17352 ft: 36520 corp: 244/2752b exec/s: 64 rss: 247Mb
#512	pulse  cov: 23370 ft: 57316 corp: 399/6397b exec/s: 56 rss: 292Mb
#1024	pulse  cov: 25687 ft: 82402 corp: 673/19Kb exec/s: 44 rss: 400Mb
fuzz: test/fuzz/tx_pool.cpp:232: auto (anonymous namespace)::tx_pool_standard_fuzz_target(FuzzBufferType)::(anonymous class)::operator()() const: Assertion `"it != result_package.m_tx_results.end()" && check' failed.
==24597== ERROR: libFuzzer: deadly signal
    #0 0x556b1cad4ab1  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x278fab1)
    #1 0x556b1ca1fc08  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26dac08)
    #2 0x556b1ca04d53  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26bfd53)
    #3 0x7f50944353bf  (/lib/x86_64-linux-gnu/libpthread.so.0+0x153bf)
    #4 0x7f509407918a  (/lib/x86_64-linux-gnu/libc.so.6+0x4618a)
    #5 0x7f5094058858  (/lib/x86_64-linux-gnu/libc.so.6+0x25858)
    #6 0x7f5094058728  (/lib/x86_64-linux-gnu/libc.so.6+0x25728)
    #7 0x7f5094069f35  (/lib/x86_64-linux-gnu/libc.so.6+0x36f35)
    #8 0x556b1cf8b787  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x2c46787)
    #9 0x556b1cb00a77  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x27bba77)
    #10 0x556b1e4256f7  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x40e06f7)
    #11 0x556b1e4253a5  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x40e03a5)
    #12 0x556b1ca06411  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c1411)
    #13 0x556b1ca05b55  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c0b55)
    #14 0x556b1ca08477  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c3477)
    #15 0x556b1ca087d9  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c37d9)
    #16 0x556b1c9f74ae  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26b24ae)
    #17 0x556b1ca202f2  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26db2f2)
    #18 0x7f509405a0b2  (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #19 0x556b1c9cc24d  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x268724d)
NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0xb3,0xb3,0xb3,0xb3,0x83,0x83,0x2f,0x2f,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xb,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0xc9,0xc9,0xcb,0x1,0x0,0xcb,
\xb3\xb3\xb3\xb3\x83\x83//\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\x0b\x00\x00\x00\x00\x00\x00\x00eeeeeeeeeee\xc9\xc9\xcb\x01\x00\xcb
artifact_prefix='./'; Test unit written to ./crash-8bc5eec8ddffef064fc1834346b11548218b6153
Base64: s7Ozs4ODLy/Ly8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLCwAAAAAAAABlZWVlZWVlZWVlZcnJywEAyw==Run tx_pool_standard with args ['/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz', '-runs=1', '/tmp/cirrus-ci-build/ci/scratch/qa-assets/fuzz_seed_corpus/tx_pool_standard']INFO: Seed: 517370066
INFO: Loaded 1 modules   (547316 inline 8-bit counters): 547316 [0x556b20760ed8, 0x556b207e68cc), 
INFO: Loaded 1 PC tables (547316 PCs): 547316 [0x556b207e68d0,0x556b21040810), 
INFO:     3949 files found in /tmp/cirrus-ci-build/ci/scratch/qa-assets/fuzz_seed_corpus/tx_pool_standard
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 3949 min: 1b max: 1048576b total: 95048838b rss: 213Mb
#128	pulse  cov: 17131 ft: 25503 corp: 122/895b exec/s: 64 rss: 225Mb
#256	pulse  cov: 17352 ft: 36520 corp: 244/2752b exec/s: 64 rss: 247Mb
#512	pulse  cov: 23370 ft: 57316 corp: 399/6397b exec/s: 56 rss: 292Mb
#1024	pulse  cov: 25687 ft: 82402 corp: 673/19Kb exec/s: 44 rss: 400Mb
fuzz: test/fuzz/tx_pool.cpp:232: auto (anonymous namespace)::tx_pool_standard_fuzz_target(FuzzBufferType)::(anonymous class)::operator()() const: Assertion `"it != result_package.m_tx_results.end()" && check' failed.
==24597== ERROR: libFuzzer: deadly signal
    #0 0x556b1cad4ab1  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x278fab1)
    #1 0x556b1ca1fc08  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26dac08)
    #2 0x556b1ca04d53  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26bfd53)
    #3 0x7f50944353bf  (/lib/x86_64-linux-gnu/libpthread.so.0+0x153bf)
    #4 0x7f509407918a  (/lib/x86_64-linux-gnu/libc.so.6+0x4618a)
    #5 0x7f5094058858  (/lib/x86_64-linux-gnu/libc.so.6+0x25858)
    #6 0x7f5094058728  (/lib/x86_64-linux-gnu/libc.so.6+0x25728)
    #7 0x7f5094069f35  (/lib/x86_64-linux-gnu/libc.so.6+0x36f35)
    #8 0x556b1cf8b787  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x2c46787)
    #9 0x556b1cb00a77  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x27bba77)
    #10 0x556b1e4256f7  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x40e06f7)
    #11 0x556b1e4253a5  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x40e03a5)
    #12 0x556b1ca06411  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c1411)
    #13 0x556b1ca05b55  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c0b55)
    #14 0x556b1ca08477  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c3477)
    #15 0x556b1ca087d9  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26c37d9)
    #16 0x556b1c9f74ae  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26b24ae)
    #17 0x556b1ca202f2  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x26db2f2)
    #18 0x7f509405a0b2  (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #19 0x556b1c9cc24d  (/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x268724d)
NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0xb3,0xb3,0xb3,0xb3,0x83,0x83,0x2f,0x2f,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xcb,0xb,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0x65,0xc9,0xc9,0xcb,0x1,0x0,0xcb,
\xb3\xb3\xb3\xb3\x83\x83//\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\x0b\x00\x00\x00\x00\x00\x00\x00eeeeeeeeeee\xc9\xc9\xcb\x01\x00\xcb
artifact_prefix='./'; Test unit written to ./crash-8bc5eec8ddffef064fc1834346b11548218b6153
Base64: s7Ozs4ODLy/Ly8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLCwAAAAAAAABlZWVlZWVlZWVlZcnJywEAyw==
```

glozow commented:
Thanks @fanquake. I've pushed a fix for what I expect is the culprit - single transaction ancestor/descendant limits are slightly looser due to CPFP carve out; the decision to not return any tx results here would have caused that crash. I've gated `CheckPackageLimits()` to only run when there's more than one transaction.

glozow commented:
Addressed review comments and added tests for size limits (thanks @ariard for motivating) Ready for review again!

ariard commented on src/validation.cpp:
```diff
@@ -1079,6 +1079,19 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
     }
 
+    // Apply package mempool ancestor/descendant limits. Skip if there is only one transaction,
+    // because it's unnecessary. Also, CPFP carve out can increase the limit for individual
```
Note, I still wonder if we need to make CPFP carve-out composable with package acceptance in the future. Otherwise a counterparty can build a branch of descendants from an output A on a multi-party unconfirmed ancestor to block package acceptance on an output B.

I don't think this is required for current LN safety, as least as long as implementation backends don't try to chain their commitment+CPFP transactions. But it might be useful for other second-layers applications (e.g Lightning Pool's batch execution+commitment txn).

JeremyRubin replied:
can we use packages to just remove the carve out?

ariard replied:
Ideally yes, carve-out don't scale for multi-party (n > 2) fee-bumping.

Though extending the carve-out to safely chain CPFPs was argued by Matt on the LDK side (see here https://lightningdevkit.slack.com/archives/CTBLT3CAU/p1625786787422100). Personally, I think this approach is a bit doomed as we can assume first-stage of the CPFP-chain can always be replaced by a valid claim of a counterparty, thus downgrading the feerate of your other broadcasted commitments and I prefer the concurrent domino fee-bumping approach even if it's more complexity swallowed by the LN backend.

Further, I think there is another multi-party contract issue we have to solve. Namely, the first-stage state transaction can be symmetric but the second-stage states asymmetric-though-composable. E.g a CTV tree with a root transaction with branch A/branch B spending 2 isolated outputs. Alice broadcast root+branch A and Bob broadcast root+branch B, thus accidentally or maliciously evicting Alice's branch. I think we would like the mempool logic to aggregate those in-flight packages by conserving the highest-feerate subgraphs and that would require some form of carve-out to avoid package limits interfering ?

ariard reviewed (commented):
ACK accf3d5. 

The conservative limits check for package acceptance as implemented by this PR sounds correct to me to comply to the package limits as currently evaluated per-transaction and also verified test coverage.

JeremyRubin commented:
utACK accf3d5

darosior reviewed (commented):
`CheckPackageLimits` (and `CheckPackage` did too) computes the virtual size without taking the number of signature operations.
It's not an issue for now as `PreChecks` (which does) is always called before any code path that result in `CheckPackageLimits` or `CheckPackage`, but we need to be careful to not introduce a DOS vector in the future.

glozow commented:
@darosior good point. I agree sigop limits should be properly enforced for packages. At the moment, since the sigops limit is checked in `PreChecks` as you said, the limit is essentially `MAX_STANDARD_TX_SIGOPS_COST` * `MAX_PACKAGE_COUNT`, or 25x the single transaction size. Should that be changed?

For `CheckPackageLimits`, I'm not sure that sigops are relevant for mempool DoS protections. We mostly care about the serialized transaction size because we don't want really large transactions to exhaust space in our mempool and we care about feerates. But we don't verify signatures after a transaction has been added to the mempool or removed from it, and don't need to know how many sigops are in our ancestors/descendants, so sigop limits probably don't make a difference?

darosior commented:
> For CheckPackageLimits, I'm not sure that sigops are relevant for mempool DoS protections

The signature ops check is not about avoiding checking too many signatures in the mempool, but to keep the mempool consistent with what could be included in the next blocks by rational miners.
Miners have a size budget (and therefore we care about feerate) but also a signature ops budget and we must take care about them too as otherwise our mempool could "diverge" (the top of the mempool wouldn't represent what would likely be included in the next block). Mixing the sig op with the feerate was a neat way to avoid the multidimensional optimization problem, since only ill-crafted transactions could get near that budget.

glozow commented:
> The signature ops check is not about avoiding checking too many signatures in the mempool, but to keep the mempool consistent with what could be included in the next blocks by rational miners.

Miners have a size budget (and therefore we care about feerate) but also a signature ops budget and we must take care about them too as otherwise our mempool could "diverge" (the top of the mempool wouldn't represent what would likely be included in the next block). Mixing the sig op with the feerate was a neat way to avoid the multidimensional optimization problem, since only ill-crafted transactions could get near that budget.

My takeaway from this is that thinking of feerate as `fee / max(sigops size, serialized size)` the same way virtual size is `max(sigops size, serialized size)` is a heuristic to address the fact that blocks are constrained by both size and sigops, which would be 2D knapsack. (I still don't think it's a DoS attack but that might just be me being nitpicky). So, what I should do is go through the validation code and make sure that the correct virtual size (including sigops) is used for checks and recorded in the mempool entries.

jnewbery commented on test/functional/mempool_package_limits.py:
```diff
@@ -0,0 +1,475 @@
+        parent_values = []
+        scripts = []
+        for _ in range(5): # Make package transactions P0 ... P4
+            gp_tx = []
```
I have a slight preference for the previous `grandparent_` naming here (I don't think `gp` will be obvious to readers). Also perhaps change this to `txs` so it's clearer that it's multiple transactions.

jnewbery commented on test/functional/mempool_package_limits.py:
```diff
@@ -0,0 +1,475 @@
+        package_hex = []
+        parent_txns = []
+        parent_values = []
+        scripts = []
```
why not `parent_scipts` to go with the other `parent_...` (or even `parent_spks`).

jnewbery commented on test/functional/mempool_package_limits.py:
```diff
@@ -0,0 +1,475 @@
+        self.test_anc_count_limits_bushy()
+
+        # The node will accept our (nonstandard) extra large OP_RETURN outputs
+        self.restart_node(0, extra_args=["-acceptnonstdtxn=1"])
```
Is it possible to make this test work without using `-acceptnonstdtxn`, perhaps by having `bulk_transaction` just add standard outputs? Since this test is all about mempool policy, it seems better to have the policy of the node being tested to be as close to realistic as possible.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -216,6 +200,80 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
+                                                      setAncestors, staged_ancestors,
+                                                      limitAncestorCount, limitAncestorSize,
+                                                      limitDescendantCount, limitDescendantSize, errString);
+    // It's possible to overestimate the ancestor/descendant totals.
```
In fact, I don't think this string ever gets used (in logging or returned to users). Should `testmempoolaccept` be updated to return the reject reason and debug message for the transaction result and package result (see `ValidationState::ToString()`, which indirectly gets called if `sendrawtransaction` fails).

jnewbery commented on test/functional/mempool_package_limits.py:
```diff
@@ -0,0 +1,475 @@
+        assert_equal(2, node.getmempoolinfo()["size"])
+        testres_too_heavy = node.testmempoolaccept(rawtxs=[pc_hex, pd_hex])
+        for txres in testres_too_heavy:
+            assert_equal(txres["package-error"], "package-mempool-limits")
```
If we improved the error messages being returned to the user, we could more precisely check that the package was rejected due to _exceeds ancestor size limits_.

jnewbery commented on test/functional/mempool_package_limits.py:
```diff
@@ -0,0 +1,475 @@
+        package_hex = []
+        for j in range(2): # Two legs (left and right)
+            # Mempool transaction (Mb and Mc)
+            mempool_tx = CTransaction()
```
No need to declare variables in Python

jnewbery commented on test/functional/mempool_package_limits.py:
```diff
@@ -0,0 +1,475 @@
+            else: # OP_TRUE
+                inputs = [{"txid": txid, "vout": 1}]
+                outputs = {self.address: value - high_fee}
+                small_tx = tx_from_hex(node.createrawtransaction(inputs, outputs))
```
Maybe choose between `tx_small` and `small_tx` and stick to it :)

jnewbery reviewed (commented):
reACK accf3d5868

I've left a bunch of small comments. None are critical, so feel free to ignore or take them in a follow-up PR.

fanquake commented:
> I've left a bunch of small comments. None are critical, so feel free to ignore or take them in a follow-up PR.

I think there's enough here to warrant a followup.

ysangkok commented on test/functional/test_framework/wallet.py:
```diff
@@ -176,3 +181,75 @@ def create_self_transfer(self, *, fee_rate=Decimal("0.003"), from_node, utxo_to_
+    signedtx = node.signrawtransactionwithkey(hexstring=rawtx, privkeys=privkeys, prevtxs=prevtxs)
+    assert signedtx["complete"]
+    tx = tx_from_hex(signedtx["hex"])
+    return (tx, signedtx["hex"], my_value, tx.vout[0].scriptPubKey.hex())
```
Users of named tuples are not forced to access them by name, because they are also tuples as the name suggests:

```
>>> A = namedtuple('A', ['x','y'])
>>> A(1,2)
A(x=1, y=2)
>>> y, x = A(1,2)
>>> x, y
(2, 1)
```

Dataclasses seem closer to what you want:

```
>>> from dataclasses import dataclass
>>> @dataclass
... class A:
...     x: int
...     y: int
... 
>>> A(1,2)
A(x=1, y=2)
>>> x,y = A(1,2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot unpack non-iterable A object
```
